staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go

```diff
@@ -148,12 +149,27 @@ func (t *envelopeTransformer) TransformToStorage(ctx context.Context, data []byt
 		return nil, err
 	}
 
+	// enforce that the last observed key ID is used by the KMS plugin
+	keyID, err := t.keyIDGetter(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	uid := string(uuid.NewUUID())
-	klog.V(6).InfoS("encrypting content using envelope service", "uid", uid, "key", string(dataCtx.AuthenticatedData()))
-	resp, err := t.envelopeService.Encrypt(ctx, uid, newKey)
+	klog.V(6).InfoS("encrypting content using envelope service",
+		"uid", uid,
+		"key", string(dataCtx.AuthenticatedData()),
+		"keyID", keyID,
+	)
+	resp, err := t.envelopeService.Encrypt(ctx, uid, keyID, newKey)
 	if err != nil {
 		return nil, fmt.Errorf("failed to encrypt DEK, error: %w", err)
 	}
+	if resp.KeyID != keyID {
+		keyIDErr := fmt.Errorf("encryption response key ID does not match observed key ID")
+		klog.ErrorS(keyIDErr, "keyID mismatch", "observed keyID", keyID, "response keyID", resp.KeyID)
+		return nil, keyIDErr
+	}
```

Comment:
@liggitt @aramase @ritazh @ibihim this proposed API change forces a KMS v2 plugin to not move "forward" in its write key (well really move in any direction) until the API server has observed it via polling of the status API (which happens every minute).  My goal with this change is to be able to assert that the key ID that was last observed by the API server will be used for new writes (instead of whatever key ID the plugin decides to use).  WDYT?

Reply:
Why would we error in this case... I expected _both_ the status poll and the encrypt response to be able to update the apiserver's view of the current write key:
* We poll status periodically so if the write key changes, we bound how long we go without realizing it
* We also look at the key id returned from encrypt responses, so we immediately know if a new write key is in use

That's how the fields in the API seem to be documented

```proto
message StatusResponse {
...
    // the current write key, used to determine staleness of data updated via value.Transformer.TransformFromStorage.
    string key_id = 3;
...
message EncryptResponse {
...
    // The KMS key ID used to encrypt the data. This must always refer to the KMS KEK and not any local KEKs that may be in use.
    // This can be used to inform staleness of data updated via value.Transformer.TransformFromStorage.
    string key_id = 2;
```